Branch for code review

sayedihashimi opened the pull request:
Using this for code review purposes, the final result will go back to the `main` branch.

MariaSolOs commented on todowebapp/src/components/TodoItem.jsx:
```diff
@@ -0,0 +1,38 @@
+﻿import PropTypes from 'prop-types';
```
I wouldn't include the use of `PropTypes` at all. They're used for checking at runtime that the types of the components match what you're expecting, but it's much better (and much more common) to use JSDoc or TypeScript to typecheck at compile time.

sayedihashimi replied:
I'm getting an error message when I don't include that

![image](https://github.com/user-attachments/assets/56cf9d20-61e3-4f92-8ff7-c8ba5749c963)

MariaSolOs replied:
Interesting. I just tried using JSDoc locally but the linting rule still complains. Since this is part of the recommended React config we can leave it as is.

sayedihashimi replied:
Ok so I will leave in the propTypes.

sayedihashimi commented on README.md:
```diff
@@ -0,0 +1,37 @@
+This is a simple TODO app built with using **[Visual Studio JavaScript and TypeScript](https://learn.microsoft.com/en-us/visualstudio/javascript/javascript-in-visual-studio?view=vs-2022)**.
```
That's how we refer to `.esproj` files, so that's why I'm using that here, but I'll change it to what you are recommending.

sayedihashimi commented on README.md:
```diff
@@ -0,0 +1,37 @@
+- **Delete Tasks**: Users can remove tasks from the list.
+- **Interactive UI**: Buttons allow task management directly from the interface.
+
+## Get started with the CLI
```
I have changed it to "## Get started outside of Visual Studio" in the next commit

sayedihashimi commented on README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Get started with the CLI
+
+Ensure that you have the following components installed.
```
I have changed it to "Ensure that you have the following tools are installed."

MariaSolOs commented on README.md:
````diff
@@ -0,0 +1,37 @@
+
+```bash
+> cd todojswebapp
+> npm install
````
Since the project contains a lock-file, `npm ci` it's better here.

sayedihashimi replied:
Thanks, I'll update that.

DanielRosenwasser commented on todowebapp/src/components/Todo.css:
```diff
@@ -0,0 +1,85 @@
+    margin-bottom: 1.25rem;
+}
+
+    #todo-input input {
```
I assume it's to make nesting more obvious, but I've never seen this practice.

FWIW, there's now https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector

sayedihashimi replied:
This is how VS indents CSS files. MadsK added this about 8 or 9 years ago. Honestly, I'm not a fan of this, but this is the default how VS behaves. Let's do some customer development to see if we should disable this by default.

sayedihashimi replied:
I will change the code to not have any nesting.

sayedihashimi commented on todowebapp/src/components/Todo.jsx:
```diff
@@ -0,0 +1,82 @@
+﻿import React, { useState } from 'react'
```
Thanks, I'll make this change on my end since I've already modified the file locally.

sayedihashimi commented:
Thanks @MariaSolOs for all your feedback. I have addressed all your comments and made changes in in the `main` branch in the commit https://github.com/sayedihashimi/todojswebapp/pull/1/commits/9ca88170136afef44a0dc760b48e41593f900b82. I'm going to close this PR it was going to an empty branch.

zkat commented on todowebapp/index.html:
```diff
@@ -0,0 +1,14 @@
+    <link href="https://fonts.googleapis.com/css?family=Inter" rel="stylesheet">
+</head>
+  <body>
+    <div id="root"></div>
```
One other thing you could do is:

```suggestion
    <main></main>
```

Thinking about it: using `<main>` inside a component, which is ostensibly meant to be encapsulated and reusable, kind of breaks that encapsulation. [A document mustn't have more than one `<main>` element that doesn't have the hidden attribute specified.](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main#:~:text=A%20document%20mustn%27t%20have%20more%20than%20one%20%3Cmain%3E%20element%20that%20doesn%27t%20have%20the%20hidden%20attribute%20specified.)

As such, we move `main` up into `index.html`, and just use `main` as our application root. 

We'll adjust our CSS slightly, since we no longer want to target `#main` with it (because of encapsulation).

Additionally, since `<main>` elements are guaranteed to be unique, we don't need the `id` anymore--it's redundant.

Related code changes are below, for your convenience.

zkat commented on todowebapp/src/main.jsx:
```diff
@@ -0,0 +1,10 @@
+import App from './App.jsx';
+import './index.css';
+
+createRoot(document.getElementById('root')).render(
```
```suggestion
createRoot(document.querySelector('main')).render(
```

This is safe: `main` elements are guaranteed to be unique.

zkat commented on todowebapp/src/components/TodoList.jsx:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    return (
+        <div id="main">
```
```suggestion
        <article class="todo-list" aria-label="task list manager">
```
I would use [`<article>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article) instead of a plain `<div>` to make it semantic. I'm also proposing changing how you target css such that it's based on classes, not unique IDs, to improve encapsulation.

zkat commented on todowebapp/src/components/TodoList.jsx:
```diff
@@ -0,0 +1,82 @@
+                    />
+                )}
+            </ol>
+        </div>
```
```suggestion
        </article>
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+#main {
```
```suggestion
.todo-list {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    max-width: 25rem;
+}
+
+h1 {
```
```suggestion
.todo-list h1 {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    color: #e0e0e0;
+}
+
+#todo-input {
```
```suggestion
.todo-input {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    margin-bottom: 1.25rem;
+}
+
+#todo-input input {
```
```suggestion
.todo-input input {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    color: #e0e0e0;
+}
+
+#todo-input .add-button {
```
```suggestion
.todo-input .add-button {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    cursor: pointer;
+}
+
+#todo-input .add-button:hover {
```
```suggestion
.todo-input .add-button:hover {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    background-color: #0056b3;
+}
+
+#todo-list {
```
```suggestion
.todo-list ol {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    border-bottom: 0.0625rem solid #333;
+}
+
+#todo-list li:last-child {
```
```suggestion
.todo-list li:last-child {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    border-bottom: none;
+}
+
+#todo-list .text {
```
```suggestion
.todo-list .text {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    flex: 1;
+}
+
+#todo-list button {
```
```suggestion
.todo-list li button {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    color: #e0e0e0;
+}
+
+#todo-list button:hover {
```
```suggestion
.todo-list li button:hover {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    color: #007bff;
+}
+
+#todo-list button.delete-button {
```
```suggestion
.todo-list li button.delete-button {
```

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    color: #ff4d4d;
+}
+
+#todo-list button.up-button,
+#todo-list button.down-button {
```
```suggestion
.todo-list li button.up-button,
.todo-list li button.down-button {
```

zkat commented on todowebapp/src/components/TodoList.jsx:
```diff
@@ -0,0 +1,82 @@
+                    Add
+                </button>
+            </div>
+            <ol id="todo-list">
```
```suggestion
            <ol aria-label="task list">
```

I've suggested moving the `todo-list` class up to the `<article>` wrapper. You can add a class to this anyway, such as `todo-list-items` if you don't want to target a plain `<ol>`.

zkat commented on todowebapp/src/components/TodoList.jsx:
```diff
@@ -0,0 +1,82 @@
+
+    return (
+        <div id="main">
+            <h1>TODO</h1>
+            <div id="todo-input">
+                <input
+                    type="text"
+                    placeholder="Enter a task"
+                    value={newTaskText}
+                    onChange={handleInputChange} />
+                <button
+                    className="add-button"
+                    onClick={addTask}>
+                    Add
+                </button>
+            </div>
```
```suggestion
            <header>
                <h1>TODO</h1>
                <div class="todo-input">
                    <input
                        type="text"
                        placeholder="Enter a task"
                        value={newTaskText}
                        onChange={handleInputChange} />
                    <button
                        className="add-button"
                        onClick={addTask}>
                        Add
                    </button>
                </div>
            </header>
```

zkat reviewed (commented):
Added some proposed changes on top of my previous ARIA suggestions: more semantic structures and elements are easier for screen readers to announce in a way that makes sense, so it's nice to avoid using divs for things that have semantic meaning, if an existing HTML element makes sense in that scope.

sayedihashimi commented:
> Added some proposed changes on top of my previous ARIA suggestions: more semantic structures and elements are easier for screen readers to announce in a way that makes sense, so it's nice to avoid using divs for things that have semantic meaning, if an existing HTML element makes sense in that scope.

Thanks @zkat! I have committed all your suggestions. I'll get it locally and test it and update the word doc.

zkat commented:
@sayedihashimi I see most of the ARIA stuff that I added directly to the word doc haven't made it over yet. Would it be helpful if I added review diffs for you so you can just click-integrate?

sayedihashimi commented:
> @sayedihashimi I see most of the ARIA stuff that I added directly to the word doc haven't made it over yet. Would it be helpful if I added review diffs for you so you can just click-integrate?

I just brought those in via https://github.com/sayedihashimi/todojswebapp/pull/2. There was a merge conflict so I had to resolve that.

zkat commented on todowebapp/src/components/TodoList.css:
```diff
@@ -0,0 +1,85 @@
+    padding: 0;
+}
+
+#todo-list li {
```
```suggestion
.todo-list li {
```

oops missed a spot, sorry!

zkat reviewed (approved):
Looking great!

sayedihashimi commented:
Closing this as this was just a review PR and all the changes have already been merged to main.
